### PR TITLE
Add a transformer-based machine translation example

### PR DESCRIPTION
**Pull request by chenmoneygithub**
Add a transformer-based machine translation example. Later on we can make this an integration test or benchmark, as the data and model are not very large.

### CODE REVIEW

File: examples/machine_translation/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+At the root directory of keras-nlp, run the following command:
+
+```shell
+python ./examples/machine_translation/train.py \
+    --num_epochs=3 \
+    --saved_model_path="saved_models/machine_translation"
+```
+
+If it finishes successfully, you should see your console prints out the
```
Comment (mattdangerw) [inline examples/machine_translation/README.md:28]: prints -> print

File: examples/machine_translation/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+```shell
+python ./examples/machine_translation/train.py \
+    --num_epochs=3 \
+    --saved_model_path="saved_models/machine_translation"
+```
+
+If it finishes successfully, you should see your console prints out the 
+following information:
+```bash
```
Comment (mattdangerw) [inline examples/machine_translation/README.md:30]: I don't think you want bash syntax highlighting. Just do a normal codeblock ```

File: examples/machine_translation/data.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    return tf.strings.regex_replace(
+        lowercase,
+        "[%s]" % re.escape(strip_chars),
+        "",
+    )
+
+
+def prepare_tokenizer(train_pairs, sequence_length, vocab_size):
+    """Preapare English and Spanish tokenizer."""
+    eng_tokenizer = TextVectorization(
```
Comment (mattdangerw) [inline examples/machine_translation/data.py:75]: perfer just a toplevel import keras.layers.TextVectorization

File: examples/machine_translation/inference.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import numpy as np
+import tensorflow as tf
+from absl import app
+from absl import flags
+from absl import logging
+
+import examples.machine_translation.data  # noqa: F401
```
Comment (mattdangerw) [inline examples/machine_translation/inference.py:20]: add comment that we are including this to import the serializable defined in the file?

File: examples/machine_translation/inference.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        "[start] Perdóname por no haberte escrito antes, por favor. [end]",
+    ),
+    ("I expected more from you.", "[start] Esperaba más de vos. [end]"),
+    ("I have a computer.", "[start] Tengo un computador. [end]"),
+    ("Dinner's ready!", "[start] ¡La cena está lista! [end]"),
+    ("Let me finish.", "[start] Déjame terminar. [end]"),
+]
+
+
+def decode_sequence(input_sentence, model, max_sequence_length, lookup_table):
```
Comment (mattdangerw) [inline examples/machine_translation/inference.py:61]: We could consider reworking this to actually use WordPiece tokenization. That would remove the need for a lot of complexity here (registering serializables, this entire function, rebuilding an inverse lookup table). You would need to download a vocabulary rather then adapt (could use the bert vocabulary I use for that example).

Might be overall more readable. Not saying we have to though, just something to consider.

File: examples/machine_translation/inference.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                decode_sequence(
+                    example[0],
+                    loaded_model,
+                    flags.sequence_length,
+                    index_lookup_table,
+                )
+            )
+
+        for i in range(len(EXAMPLES)):
+            print("ENGLISH SEnTENCE: ", EXAMPLES[i][0])
```
Comment (mattdangerw) [inline examples/machine_translation/inference.py:113]: SEnTENCE -> SENTENCE

File: examples/machine_translation/train.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+flags.DEFINE_string(
+    "saved_model_path",
+    "saved_models/machine_translation_model",
+    "The path to saved model",
+)
+
+
+def train_loop(model, train_ds, val_ds):
```
Comment (mattdangerw) [inline examples/machine_translation/train.py:59]: Is there a reason we can't use `fit` here?

File: keras_nlp/integration_tests/import_test.py [hunk 8]
```diff
@@ -1,20 +0,0 @@
...
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
-import keras_nlp
-
-
-# Basic test to make sure the library can be imported.
-def test_import():
```
Comment (mattdangerw) [inline keras_nlp/integration_tests/import_test.py:19]: Why remove this? This is the only auto run test that makes sure our __init__.py doesn't error out. And it doesn't seem like you are replacing it here.

**Review by mattdangerw — COMMENTED** [review]
Left some initial comments!

File: examples/machine_translation/README.md (hunk 1 above)
Comment (chenmoneygithub) [reply examples/machine_translation/README.md]: done

File: examples/machine_translation/README.md (hunk 2 above)
Comment (chenmoneygithub) [reply examples/machine_translation/README.md]: Thanks, done!

File: examples/machine_translation/data.py (hunk 3 above)
Comment (chenmoneygithub) [reply examples/machine_translation/data.py]: done

File: examples/machine_translation/inference.py (hunk 4 above)
Comment (chenmoneygithub) [reply examples/machine_translation/inference.py]: done

File: examples/machine_translation/inference.py (hunk 5 above)
Comment (chenmoneygithub) [reply examples/machine_translation/inference.py:63]: This is a good point, I will create an issue and do it in a separate PR. 

One question - does wordpiece tokenization support this autoregressive text generation?

File: examples/machine_translation/inference.py (hunk 6 above)
Comment (chenmoneygithub) [reply examples/machine_translation/inference.py]: done.

File: examples/machine_translation/train.py (hunk 7 above)
Comment (chenmoneygithub) [reply examples/machine_translation/train.py]: Somehow I thought model.fit() does not allow masking, but actually we can use `sample_weights`. 

Changed to model.fit().

File: keras_nlp/integration_tests/import_test.py (hunk 8 above)
Comment (chenmoneygithub) [reply keras_nlp/integration_tests/import_test.py]: this is a mistake... I have added them back

File: examples/machine_translation/inference.py (hunk 5 above)
Comment (mattdangerw) [reply examples/machine_translation/inference.py:63]: I think the usually approach would be to do the "autoregressive" part in token id space (keep a running list of token ids). And then detokenize a single sample/paragraph/etc at a time.

You could detokenize one token at a time if you wanted, but it would be less efficient, and you would probably need to check for the suffix indicator in each piece yourself.

Comment (mattdangerw) [reply examples/machine_translation/inference.py:63]: Follow up sounds good to me!

File: examples/machine_translation/train.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer("num_epochs", 1, "Number of epochs to train.")
+flags.DEFINE_integer("steps_per_epoch", None, "Number of steps per epoch.")
+flags.DEFINE_integer("num_encoders", 2, "Number of Transformer encoder layers.")
+flags.DEFINE_integer("num_decoders", 2, "Number of Transformer decoder layers.")
+flags.DEFINE_integer("batch_size", 64, "The training batch size.")
+flags.DEFINE_float("learning_rate", 0.001, "The initial learning rate.")
+flags.DEFINE_integer("embed_dim", 64, "Embedding size.")
```
Comment (mattdangerw) [inline examples/machine_translation/train.py:30]: Should we rename this model_dim? model_hidden_dim? encoding_size?

I initially read this as the size of the embedding only, had to check rest of code to confirm this is also the main dimension for the encoder/decoder blocks.

Could alternately just leave a comment somewhere explaining.

File: examples/machine_translation/data.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+                "decoder_inputs": spa[:, :-1],
+            },
+            spa[:, 1:],
+            tf.cast((spa[:, 1:] != 0), tf.float32),  # mask as sample weights
+        )
+
+    dataset = tf.data.Dataset.from_tensor_slices((eng_texts, spa_texts))
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(format_dataset)
+    return dataset.shuffle(2048).prefetch(16).cache()
```
Comment (mattdangerw) [inline examples/machine_translation/data.py:119]: can we prefetch(tf.data.AUTOTUNE) just to cut down on magic numbers in this example?

File: examples/machine_translation/train.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+from absl import app
+from absl import flags
+
+from examples.machine_translation.data import get_dataset_and_tokenizer
+from examples.machine_translation.model import TranslationModel
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer("num_epochs", 1, "Number of epochs to train.")
+flags.DEFINE_integer("steps_per_epoch", None, "Number of steps per epoch.")
```
Comment (mattdangerw) [inline examples/machine_translation/train.py:25]: These are finite sized datasets right? Why do we need this?

**Review by mattdangerw — APPROVED** [review]
Looks good to me! Left a few more comments.

File: examples/machine_translation/train.py (hunk 11 above)
Comment (chenmoneygithub) [reply examples/machine_translation/train.py:25]: ohh, this is useful if users do not want to run through the whole datasets. Particularly useful for debugging.

File: examples/machine_translation/train.py (hunk 9 above)
Comment (chenmoneygithub) [reply examples/machine_translation/train.py:30]: sg, I am renaming it to model_dim.

File: examples/machine_translation/data.py (hunk 10 above)
Comment (chenmoneygithub) [reply examples/machine_translation/data.py:119]: sg! changed.

File: examples/machine_translation/data.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+def custom_standardization(input_string):
+    lowercase = tf.strings.lower(input_string)
+    return tf.strings.regex_replace(
+        lowercase,
+        "[%s]" % re.escape(strip_chars),
+        "",
+    )
+
+
+def prepare_tokenizer(train_pairs, sequence_length, vocab_size):
```
Comment (fchollet) [inline examples/machine_translation/data.py:72]: tokenizers (plural)

File: examples/machine_translation/model.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        positions = tf.range(start=0, limit=length, delta=1)
+        embedded_tokens = self.token_embeddings(inputs)
+        embedded_positions = self.position_embeddings(positions)
+        return embedded_tokens + embedded_positions
+
+    def compute_mask(self, inputs, mask=None):
+        return tf.math.not_equal(inputs, 0)
+
+
+class TranslationModel(keras.Model):
```
Comment (fchollet) [inline examples/machine_translation/model.py:47]: This model simply chains layers together, it contains no custom logic. So it would be more concise if defined as a Functional API model.

something like

```
def get_translation_model(num_encoders, num_decoders, num_heads, ...):
   return Model(...)
```

File: examples/machine_translation/model.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.layers import TransformerDecoder
+from keras_nlp.layers import TransformerEncoder
+
+
+class PositionalEmbedding(keras.layers.Layer):
```
Comment (fchollet) [inline examples/machine_translation/model.py:21]: We should directly add this class to the API. It is very simply. We just need to settle on the signature (it should match the signature of the `Embedding` layer).

File: examples/machine_translation/data.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+                "decoder_inputs": spa[:, :-1],
+            },
+            spa[:, 1:],
+            tf.cast((spa[:, 1:] != 0), tf.float32),  # mask as sample weights
+        )
+
+    dataset = tf.data.Dataset.from_tensor_slices((eng_texts, spa_texts))
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(format_dataset)
+    return dataset.shuffle(2048).prefetch(tf.data.AUTOTUNE).cache()
```
Comment (fchollet) [inline examples/machine_translation/data.py:119]: shuffle buffer should be batch size dependent. And should happen before batching.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: examples/machine_translation/model.py (hunk 13 above)
Comment (chenmoneygithub) [reply examples/machine_translation/model.py:47]: I am thinking for all examples we will add, maybe keep a uniform style? Also subclassing approach is easier to scale, if later we want to add something fancy to this example. WDYT?

File: examples/machine_translation/model.py (hunk 14 above)
Comment (chenmoneygithub) [reply examples/machine_translation/model.py:21]: I believe Fabien is looking at the design of positional embedding, so let's now keep this in the example, and later replace it?

File: examples/machine_translation/data.py (hunk 12 above)
Comment (chenmoneygithub) [reply examples/machine_translation/data.py:72]: done

File: examples/machine_translation/data.py (hunk 15 above)
Comment (chenmoneygithub) [reply examples/machine_translation/data.py:119]: done
